### PR TITLE
Improve default dynamic CP packing scheduler

### PR DESCRIPTION
**Pull request by xiaoyao0115**
## Summary
- Replace the default dynamic CP scheduler path with packing-aware CP group selection
- Allow shorter sequences to use larger CP groups when that lets them pack with longer sequences and lowers the critical-path workload
- Add focused coverage for the packing-aware grouping behavior

## Testing
- `git diff --check upstream/dev`
- `python -m compileall -q megatron/core/datasets/data_schedule.py megatron/core/datasets/data_schedule_utils.py megatron/core/model_parallel_config.py tests/unit_tests/test_sequence_packing.py`

Not run: full pytest, because the local shell Python does not have torch/pytest available.

### CODE REVIEW

**Review by kunlunl — APPROVED** [review]

File: megatron/core/datasets/data_schedule_utils.py [hunk 1]
```diff
@@ -561,241 +557,179 @@ def next_hdp_group(
...
+                new_sample_ids_per_gpu[rank] = sample_ids_per_gpu[rank]
 
-        min_group_size = min(existing_group_sizes)
-        next_power = min(min_group_size * 2, total_gpus)
+            for rank in range(group_start_rank, group_end_rank + needed_count + 1):
+                new_micro_batches[rank] = micro_batches[group_end_rank]
+                new_exec_times[rank] = sum(
+                    workload(length, next_power) for length in micro_batches[group_end_rank]
+                )
+                new_sample_ids_per_gpu[rank] = sample_ids_per_gpu[group_end_rank]
```
[the remark this answers is not in the document]
Comment (xiaoyao0115) [reply megatron/core/datasets/data_schedule_utils.py]: fixed

File: megatron/core/datasets/data_schedule_utils.py [hunk 2]
```diff
@@ -561,241 +557,179 @@ def next_hdp_group(
...
+                chosen_members = sorted(free_ranks, key=lambda rank: exec_times[rank])[:cp_size]
+                chosen_set = set(chosen_members)
+                projected_max = max(
+                    time + per_gpu_cost if rank in chosen_set else time
+                    for rank, time in enumerate(exec_times)
+                )
+                if projected_max <= cap and (best is None or projected_max < best[0]):
+                    best = (projected_max, cp_size, "new", None, chosen_members)
+
+            cp_size *= 2
```
[the remark this answers is not in the document]
Comment (xiaoyao0115) [reply megatron/core/datasets/data_schedule_utils.py:639]: fixed

File: megatron/core/datasets/data_schedule_utils.py [hunk 3]
```diff
@@ -561,241 +557,179 @@ def next_hdp_group(
...
+                continue
 
-        if needed < prev_needed:
-            check_balance = True
+            members = group_members[group_id]
+            needed_count = next_power - min_group_size
+            group_start_rank = members[0]
+            group_end_rank = members[-1]
+            empty_rank = empty_ranks[0]
+            if group_end_rank + 1 > empty_rank or group_end_rank + needed_count >= total_gpus:
```
[the remark this answers is not in the document]
Comment (xiaoyao0115) [reply megatron/core/datasets/data_schedule_utils.py:691]: fixed

File: megatron/core/datasets/data_schedule_utils.py [hunk 4]
```diff
@@ -561,241 +557,179 @@ def next_hdp_group(
...
-                sample_seq_tuple, bucket_idx = cand_tuple, idx
-                break
+    def workload(seq_len: int, cp_size: int) -> float:
+        return (seq_len * seq_len) / cp_size
 
-        if sample_seq_tuple is None:
-            break
+    sample_seqlens = sorted(sample_seqlens, key=lambda x: x[1], reverse=True)
+    local_tall = sample_seqlens[0][1]
+    cap = float(local_tall) * float(max_seq_len_per_rank) * (1.0 + _DYNAMIC_CP_WORKLOAD_CAP_DELTA)
```
[the remark this answers is not in the document]
Comment (xiaoyao0115) [reply megatron/core/datasets/data_schedule_utils.py:570]: fixed

File: tests/unit_tests/test_sequence_packing.py [hunk 5]
```diff
@@ -611,21 +638,15 @@ def _check_batch(batch_all, batch_keys):
...
-                        mb_cp_group = parallel_state.get_dynamic_data_context_parallel_groups(
-                            group_size=local_cp
-                        )
-                        torch.distributed.all_reduce(
-                            mb_sum, op=torch.distributed.ReduceOp.SUM, group=mb_cp_group
-                        )
-                        # all_reduce result = mb_sum * local_cp.
-                        # Scale to mb_sum * max_cp.
                         mb_sum *= max_cp // local_cp
                         token_sum_after += mb_sum
```
[the remark this answers is not in the document]
Comment (xiaoyao0115) [reply tests/unit_tests/test_sequence_packing.py]: fixed

File: megatron/core/datasets/data_schedule_utils.py [hunk 6]
```diff
@@ -561,241 +557,186 @@ def next_hdp_group(
...
-            micro_batches, exec_times, sample_ids_per_gpu, group_members, group_size
-        )
-        empty_gpus = any([not micro_batches[i] for i in range(total_gpus)])
+        return False
 
-    total_work_after = sum(len(mb) for mb in micro_batches)
-    assert (
-        total_work_after >= total_work_before
-    ), f"Samples were removed: {total_work_before} -> {total_work_after}"
+    while any(not micro_batch for micro_batch in micro_batches):
```
Comment (Victarry) [inline megatron/core/datasets/data_schedule_utils.py:768]: **[IMPORTANT Correctness]** `next_hdp_group_packing_aware()` can return microbatches with empty DPxCP ranks.

The function documents the “no empty DPxCP rank” invariant, but the fill loop silently exits when `fill_empty_gpus_once()` cannot expand another group:

```python
while any(not micro_batch for micro_batch in micro_batches):
    if not fill_empty_gpus_once():
        break
```

Downstream `build_packed_microbatches()` assumes every rank has at least one sample and indexes `sample_ids_this_group[0]`, so this becomes an `IndexError` or rank-divergent scheduling failure. A source-equivalent simulation of this algorithm leaves empty ranks for non-power-of-two DPxCP sizes, e.g. `total_gpus=14`, `max_seq_len_per_rank=100`, `sample_seqlens=[(0, 50), (1, 50)]`.

**Suggestion:** Do not return partial groups. Either make full `total_gpus` a candidate group size for non-power-of-two DPxCP layouts and keep expanding until no rank is empty, or assert/raise here with a clear error before returning. Please also add a unit test with a non-power-of-two DPxCP size.

File: tests/unit_tests/test_sequence_packing.py [hunk 7]
```diff
@@ -611,21 +638,15 @@ def _check_batch(batch_all, batch_keys):
                 # After wrap.
                 token_sum_after = torch.tensor(0, dtype=torch.int64, device='cuda')
                 if is_dynamic_cp:
-                    # DCP: per-microbatch CP all_reduce + scale to max_cp,
-                    # then dp_cp all_reduce to aggregate unique contributions.
+                    # DCP: each logical sequence is replicated on local_cp ranks
+                    # before CP slicing. Scale each rank's local contribution by
+                    # max_cp / local_cp, then reduce across DPxCP ranks so each
+                    # sequence contributes exactly max_cp times.
                     for batch in batch_all:
```
Comment (Victarry) [inline tests/unit_tests/test_sequence_packing.py:654]: **[IMPORTANT Test Correctness]** The DCP token-sum check no longer reconstructs each local CP group before scaling.

`get_cp_slice_for_thd()` slices `tokens` per CP rank before this test reads `batch['tokens']`. Without the local-CP `all_reduce`, each rank contributes only its CP slice. The final DPxCP all-reduce then computes `full_sequence_sum * max_cp / local_cp` instead of `full_sequence_sum * max_cp` for `local_cp > 1`, which weakens/removes the invariant this test was checking and can hide token-loss bugs in DCP packing.

**Suggestion:** Restore the per-microbatch local CP all-reduce before scaling:

```python
mb_cp_group = parallel_state.get_dynamic_data_context_parallel_groups(group_size=local_cp)
torch.distributed.all_reduce(mb_sum, op=torch.distributed.ReduceOp.SUM, group=mb_cp_group)
mb_sum *= max_cp // local_cp
```

Then reduce `token_sum_after` across `dp_cp_group` as before.

**Review by Victarry — APPROVED** [review]

File: tests/unit_tests/test_sequence_packing.py (hunk 7 above)
Comment (xiaoyao0115) [reply tests/unit_tests/test_sequence_packing.py:654]: I think this check is still running before CP slicing.

In this test, `batch_all` is read directly from the iterator returned by `wrap_data_iterator()`.

File: megatron/core/datasets/data_schedule_utils.py (hunk 6 above)
Comment (xiaoyao0115) [reply megatron/core/datasets/data_schedule_utils.py:768]: fixed
